packages/core/snapshot/write.go

```diff
@@ -158,26 +126,124 @@ func NewLedgerUTXOStatesProducer(nmgr *notarization.Manager) UTXOStatesProducerF
 
 // NewEpochDiffsProducer returns a OutputWithMetadataProducerFunc that provide OutputWithMetadatas from the ledger.
 func NewEpochDiffsProducer(fullEpochIndex, latestCommitableEpoch epoch.Index, nmgr *notarization.Manager) EpochDiffProducerFunc {
-	epochDiffs, err := nmgr.SnapshotEpochDiffs(fullEpochIndex, latestCommitableEpoch)
+	prodChan := make(chan *ledger.EpochDiff)
+	stopChan := make(chan struct{})
+	nmgr.SnapshotEpochDiffs(fullEpochIndex, latestCommitableEpoch, prodChan, stopChan)
 
-	return func() (map[epoch.Index]*ledger.EpochDiff, error) {
-		return epochDiffs, err
+	return func() *ledger.EpochDiff {
+		select {
+		case obj := <-prodChan:
+			return obj
+		case <-stopChan:
+			close(prodChan)
+			return nil
+		}
 	}
 }
 
+func writeEpochDiffs(writeSeeker io.WriteSeeker, diffs *ledger.EpochDiff) error {
+	writeFunc := func(name string, value any) error {
+		return writeFunc(writeSeeker, name, value)
+	}
+
+	spentLen := len(diffs.Spent())
+	if err := writeFunc("epochDiffs spent Len", int64(spentLen)); err != nil {
+		return err
+	}
+
+	s := diffs.Spent()
+	var end int
+	for i := 0; i < spentLen; {
+		if i+chunkSize > spentLen {
+			end = spentLen
+		} else {
+			end = i + chunkSize
+		}
+		writeOutputsWithMetadata(writeSeeker, s[i:end])
+		i = end
+	}
+
+	createdLen := len(diffs.Created())
+	if err := writeFunc("epochDiffs created Len", int64(createdLen)); err != nil {
+		return err
+	}
+	c := diffs.Created()
+	for i := 0; i < createdLen; {
+		if i+chunkSize > createdLen {
+			end = createdLen
+		} else {
+			end = i + chunkSize
+		}
+		writeOutputsWithMetadata(writeSeeker, c[i:end])
+		i = end
+	}
+
+	return nil
+}
+
 func writeSolidEntryPoints(writeSeeker io.WriteSeeker, seps *SolidEntryPoints) error {
 	writeFunc := func(name string, value any) error {
 		return writeFunc(writeSeeker, name, value)
 	}
 
-	data, err := serix.DefaultAPI.Encode(context.Background(), seps, serix.WithValidation())
-	if err != nil {
+	// write EI
+	if err := writeFunc("solid entry points epoch", seps.EI); err != nil {
 		return err
 	}
-	if err := writeFunc("seps", append(data, delimiter...)); err != nil {
+
+	// write number of solid entry points
+	sepsLen := len(seps.Seps)
+	if err := writeFunc("solid entry points Len", int64(sepsLen)); err != nil {
 		return err
 	}
 
+	// write solid entry points in chunks
+	s := seps.Seps
+	var end int
+	for i := 0; i < sepsLen; {
+		if i+chunkSize > sepsLen {
+			end = sepsLen
+		} else {
+			end = i + chunkSize
+		}
+
+		data, err := serix.DefaultAPI.Encode(context.Background(), s[i:end], serix.WithValidation())
+		if err != nil {
+			return err
+		}
+
+		if err := writeFunc("sepsBytesLen", int64(len(data))); err != nil {
+			return err
+		}
+		if err := writeFunc("seps", data); err != nil {
+			return err
+		}
+
+		i = end
+	}
+
+	return nil
+}
+
+func writeOutputsWithMetadata(writeSeeker io.WriteSeeker, outputsChunks []*ledger.OutputWithMetadata) error {
+	if len(outputsChunks) == 0 {
```

Comment:
Are these outputsChunks? This should be just one chunk of outputs no?

Reply:
No, it writes a slice of OutputWithMetadata at once.